### ISSUE TITLE
Nim crashes randomly

### ISSUE THREAD

**Issue by yglukhov** [nim-lang/Nim#16613]
**Steps to reproduce**:
```
git clone https://github.com/yglukhov/iface
cd iface
git checkout c137f9330cc0bb59b1c9204fecd2358b5c156f52
while nim c tests/test1.nim; do true; done
```
Wait for 5 minutes
**Actual result**:
Nim eventually crashes with stack trace:
```
Traceback (most recent call last)
.../nim-devel/compiler/nim.nim(128) nim
.../nim-devel/compiler/nim.nim(87) handleCmdLine
.../nim-devel/compiler/main.nim(275) mainCommand
.../nim-devel/compiler/main.nim(245) compileToBackend
.../nim-devel/compiler/main.nim(101) commandCompileToC
.../nim-devel/compiler/modules.nim(178) compileProject
.../nim-devel/compiler/modules.nim(98) compileModule
.../nim-devel/compiler/passes.nim(180) processModule
.../nim-devel/compiler/passes.nim(73) processTopLevelStmt
.../nim-devel/compiler/sem.nim(644) myProcess
.../nim-devel/compiler/sem.nim(612) semStmtAndGenerateGenerics
.../nim-devel/compiler/semstmts.nim(2335) semStmt
.../nim-devel/compiler/semexprs.nim(1055) semExprNoType
.../nim-devel/compiler/semexprs.nim(2970) semExpr
.../nim-devel/compiler/semstmts.nim(2277) semStmtList
.../nim-devel/compiler/semexprs.nim(2852) semExpr
.../nim-devel/compiler/semexprs.nim(1038) semDirectOp
.../nim-devel/compiler/semexprs.nim(921) afterCallActions
.../nim-devel/compiler/semexprs.nim(39) semTemplateExpr
.../nim-devel/compiler/sem.nim(421) semAfterMacroCall
.../nim-devel/compiler/semstmts.nim(2335) semStmt
.../nim-devel/compiler/semexprs.nim(1055) semExprNoType
.../nim-devel/compiler/semexprs.nim(2970) semExpr
.../nim-devel/compiler/semstmts.nim(2277) semStmtList
.../nim-devel/compiler/semexprs.nim(2969) semExpr
.../nim-devel/compiler/semexprs.nim(2552) semBlock
.../nim-devel/compiler/semexprs.nim(2970) semExpr
.../nim-devel/compiler/semstmts.nim(2277) semStmtList
.../nim-devel/compiler/semexprs.nim(2978) semExpr
.../nim-devel/compiler/semstmts.nim(204) semTry
.../nim-devel/compiler/semstmts.nim(109) semExprBranchScope
.../nim-devel/compiler/semstmts.nim(102) semExprBranch
.../nim-devel/compiler/semexprs.nim(2970) semExpr
.../nim-devel/compiler/semstmts.nim(2277) semStmtList
.../nim-devel/compiler/semexprs.nim(2970) semExpr
.../nim-devel/compiler/semstmts.nim(2277) semStmtList
.../nim-devel/compiler/semexprs.nim(2852) semExpr
.../nim-devel/compiler/semexprs.nim(1038) semDirectOp
.../nim-devel/compiler/semexprs.nim(921) afterCallActions
.../nim-devel/compiler/semexprs.nim(39) semTemplateExpr
.../nim-devel/compiler/sem.nim(421) semAfterMacroCall
.../nim-devel/compiler/semstmts.nim(2335) semStmt
.../nim-devel/compiler/semexprs.nim(1055) semExprNoType
.../nim-devel/compiler/semexprs.nim(2970) semExpr
.../nim-devel/compiler/semstmts.nim(2277) semStmtList
.../nim-devel/compiler/semexprs.nim(2948) semExpr
.../nim-devel/compiler/semstmts.nim(162) semIf
.../nim-devel/compiler/semstmts.nim(102) semExprBranch
.../nim-devel/compiler/semexprs.nim(2970) semExpr
.../nim-devel/compiler/semstmts.nim(2277) semStmtList
.../nim-devel/compiler/semexprs.nim(2978) semExpr
.../nim-devel/compiler/semstmts.nim(204) semTry
.../nim-devel/compiler/semstmts.nim(109) semExprBranchScope
.../nim-devel/compiler/semstmts.nim(102) semExprBranch
.../nim-devel/compiler/semexprs.nim(2970) semExpr
.../nim-devel/compiler/semstmts.nim(2277) semStmtList
.../nim-devel/compiler/semexprs.nim(2970) semExpr
.../nim-devel/compiler/semstmts.nim(2277) semStmtList
.../nim-devel/compiler/semexprs.nim(2974) semExpr
.../nim-devel/compiler/semstmts.nim(675) semConst
.../nim-devel/compiler/semexprs.nim(86) semExprWithType
.../nim-devel/compiler/semexprs.nim(70) semExprCheck
.../nim-devel/compiler/semexprs.nim(2857) semExpr
.../nim-devel/compiler/semexprs.nim(276) semConv
.../nim-devel/compiler/semexprs.nim(851) semStaticExpr
.../nim-devel/compiler/vm.nim(2220) evalStaticExpr
.../nim-devel/compiler/vm.nim(2212) evalConstExprAux
.../nim-devel/compiler/vm.nim(815) rawExecute
SIGSEGV: Illegal storage access. (Attempt to read from nil?)
```
Nim version: bbe05c15326c7a62d15c0fe47d1c6e1be3ab11a8

**Comment by timotheecour** [comment]
@yglukhov can you:
* always report nim version (git hash) in bug reports (in particular line numbers are not meaningful without a version)
* minimize the example so it contains as few dependencies as possible, ideally only stdlib modules and preferably a single module without imports

to help narrow down, several flags can help:
* building nim with `--stacktrace --stacktraceMsgs -d:nimCompilerStacktraceHints` so that you know where you are in user code in the stacktrace
* instrumenting code in the place it crashes (maybe `of opcWrDeref:`) to show the line info in user code right before crash (info is `c.debug[pc]`)
* `--processing:filenames`
* vmutils.vmTrace to trace VM execution in user code

EDIT: it's not impossible this is related to https://github.com/nim-lang/Nim/issues/17527, and that it'd be caused by a GC issue, can you read this comment https://github.com/nim-lang/Nim/issues/17527#issuecomment-808636909 and try with `-d:useSysAssert -d:useGcAssert` ?

**Comment by yglukhov** [comment]
* I've updated description with the most recent nim version and its stack trace.
* `--stacktrace --stacktraceMsgs -d:nimCompilerStacktraceHints -d:useSysAssert -d:useGcAssert` doesn't change the stack trace of the crash
* logging `c.debug[pc]` seems to point to this line: https://github.com/yglukhov/iface/blob/c137f9330cc0bb59b1c9204fecd2358b5c156f52/tests/test1.nim#L32

@timotheecour were you unable to reproduce the crash?

**Comment by timotheecour** [comment]
> doesn't change the stack trace of the crash

that's odd, it should; did you re-build nim with those flags (which is what's needed) or compile your program with those flags?

next step would be to minimize to remove any module dependency, since it's a compiler bug

**Comment by yglukhov** [comment]
> next step would be to minimize to remove any module dependency, since it's a compiler bug

Does it mean that you tried to reproduce it and failed?

**Comment by timotheecour** [comment]
with `--stacktrace --stacktraceMsgs -d:nimCompilerStacktraceHints` i get as of 7d3c3e00efe777dcbda15f2125fa77aa7d17cad4: https://gist.github.com/timotheecour/7f04794b8bc18a05e6e1047c6ae0b95f#file-gistfile1-txt-L2

which does show the stacktrace messages
you probably didn't compile nim with those flags and used those flags to compile the main program instead

(it crashed after running for 2 minutes)

> Does it mean that you tried to reproduce it and failed?

it means it's hard to investigate without the bug minimized

**Comment by timotheecour** [comment]
TLDR: see workaround below, this would be fixed by https://github.com/nim-lang/RFCs/issues/276

* crashes in 1.0, 1.2, 1.4 also
* doesn't compile in 0.19: iface.nim(63, 3) Error: undeclared identifier: 'signatureHash'
* after instrumentation below it fails because `n2 == nil:`
```nim
        if regs[ra].nodeAddr == nil:
          stackTrace(c, tos, pc, "D20210729T134630")
        let n2 = regs[ra].nodeAddr[]
        if n2 == nil:
          stackTrace(c, tos, pc, "D20210729T134630.2")
        if n == nil:
          stackTrace(c, tos, pc, "D20210729T134630.3")
        if (nfIsRef notin regs[ra].nodeAddr[].flags and
            nfIsRef notin n.flags):
          regs[ra].nodeAddr[][] = n[]
```
* vmutils.vmtrace shows it fails here:
```
/Users/timothee/git_clone/nim/temp/iface/tests/test1.nim(43, 23) [opcAsgnComplex]       let res = doSmth(d)
```
* possibly related to converter because of the iface code which generates this: https://gist.github.com/timotheecour/5fc8f646d09cf32a85ed820723a41f72

* slightly reduced:
```nim
import iface, unittest
# import std/vmutils

iface Animal:
  proc say(): string
  proc test2()

type
  Dog = ref object of RootRef
    testCalled: bool

proc say(d: Dog): string = "bark"
proc test2(d: Dog) =
  d.testCalled = true

proc doSmth(a: Animal): string =
  a.test2()
  result = a.say()

proc createVoldemortObject(): Animal =
  type Hidden = ref object of RootRef
  proc test2(h: Hidden) = discard

suite "iface":
  test "runtime":
    let d = Dog.new()
    doAssert doSmth(d) == "bark"
    doAssert d.testCalled

  block:
    # static: vmTrace(true)
    when defined case1:
      const s = static:
        let d = Dog.new()
        let res = doSmth(d)
        doAssert(d.testCalled)
        0
    when defined case2:
      const s = (proc(): auto =
        let d = Dog.new()
        let res = doSmth(d)
        doAssert(d.testCalled)
        0)()
```
=> crashes with -d:case1
=> works with -d:case2, aka https://github.com/nim-lang/RFCs/issues/276

* would be fixed by https://github.com/nim-lang/RFCs/issues/276, which also gives you a workaround until that RFC is implemented: either replace the static block by a call to an auxiliary proc, or use an immediately invoked lambda:

```nim
    # const s = static:
    const s = (proc(): auto =
      let d = Dog.new()
      let res = doSmth(d)
      doAssert(d.testCalled)
      0)()
```
=> no crash

regardless of https://github.com/nim-lang/RFCs/issues/276, this should be further investigated, ideally by finding a reduced example that exhibits the bug all the time instead of just once every 2 minutes

**Comment by yglukhov** [comment]
The crash happens, because PNode gets collected as no refs point to it. It is accessed though `nodeAddr` of registry. The value of the collected Node address is set here:
https://github.com/nim-lang/Nim/blob/bbe05c15326c7a62d15c0fe47d1c6e1be3ab11a8/compiler/vm.nim#L680

**Comment by timotheecour** [comment]
> The crash happens, because PNode gets collected as no refs point to it. It is accessed though nodeAddr of registry. The value of the collected Node address is set here:

possibly, but can you show a reduced test case (with no imports) to demonstrate this? Once we have a reduced test case, we can think about ways to fix this, possibly by adding gc'd references to all PNode's for which we've taken the address, storing in some `seq[PNode]`

**Comment by yglukhov** [comment]
> possibly, but can you show a reduced test case (with no imports) to demonstrate this?

I don't have any, and I'm not sure it'll help, judging from the randomness of this issue. However, I've added some [code to nim](https://github.com/yglukhov/Nim/pull/new/debug-16613) that proves my point and shows 100% reproducibility.

**Comment by timotheecour** [comment]
that's great to have a way to reproduce at 100%, I can indeed reproduce via https://github.com/nim-lang/Nim/pull/18628

However, we still need a minimized test case (now that we have a way to reproduce at 100% via that PR), to make it easier to understand the issue and fix it (even if it's indeed due to gc collecting nodeAddr)

A potential fix is by maintaining a `seq[PNode]` in the compiler as i suggested above, but it also raises the question of whether it'd prevent collecting things that should be collected; again, a minimized test case would help.

**Comment by yglukhov** [comment]
> A potential fix is by maintaining a seq[PNode] in the compiler as i suggested above, but it also raises the question of whether it'd prevent collecting things that should be collected; again, a minimized test case would help.

I've been thinking of a similar approach, but not to use a `seq[PNode]`, but just a `gcHold: PNode` field next to `nodeAddr` field, which should always be set along with `nodeAddr` to whatever is needed.

**Comment by yglukhov** [comment]
Ideally though it would be great to get rid of unsafe code (`ptr` that is) entirely. Not sure how feasible is that.

**Comment by yglukhov** [comment]
#19515 looks a lot like it could fix this issue, doesn't it?
